[bcs] generics + bcs schema

damirka opened the pull request:
So far it is a draft and the PR is lacking documentation. Quick summary of changes:

```js 
// bcs is now an instance
// address type is built-in based on the configuration
// generics are now supported out of the box for vector and all of the types

import { BCS } from "@mysten/bcs";

const bcs = new BCS({
  vectorType: 'vector', // name of the vector type
  addressLength: 20, // address of the address type (eg 20 Sui, 32 Aptos)
  addressEncoding: 'hex',
  genericSeparators: ['<', '>'] // open and close symbols
});

bcs.registerStructType('Container<T>', {
  owner: 'address',
  content: 'T'
});

let bytes = bcs.ser('Container<u64>', { owner: '0x.......', content: '1337' }).toString('hex');

// same for de

bcs.de('Container<u64>', bytes, 'hex');

// also vectors

bcs.ser('vector<Container<u64>>', [ { owner: '...', content: '1337 } ] /* .... */).toString('hex'); 
```

amnn commented on sdk/bcs/src/index.ts:
```diff
@@ -371,17 +371,36 @@ function ulebDecode(
+/**
+ *
+ */
+export type TypeSchema = {
```
Is `Schema` the established name for what this is?  I imagine that it's quite an overloaded word in the BCS world, e.g. when I read `TypeSchema` I understand it to be similar to what `TypeInterface` is right now.  The content of this struct suggests to me a name like `BCSConfig` or maybe `TypeGrammar`?

damirka replied:
I was planning to have two groups of type definitions:
```json
{
  "structs": {
    "Coin<T>": {
      "value": "u64"
    }
   },
   "enums": {
     "Option<T>": {
       "none": null,
       "some": "T"
     }
   }
}
```

amnn reviewed (commented):
Love the change from singleton to allowing multiple instances!  I'm also a fan of @Jordan-Mysten's suggestion of a `fork` API.

Because there are a couple of other changes in this PR that I have less context on, I'm holding off on accept so other folks can weigh in on those.

On a meta-note, it would be easier to review these changes if they were factored out each into their own PR (I would have been able to stamp the PR that moved the interface away from the singleton pattern for example).  I don't know the best way to do that in Github unfortunately, maybe someone who is more familiar with its workflows can chime in!  I'm more familiar with phabricator where we had the notion of "stacked" PRs where each change could be reviewed separately but you knew they were related.  But here are the potentially separate PRs I can see included in this one:

- Singleton to individual instances
- Generics/type parameters
- Per-language configuration of the type grammar

oxade commented on sdk/bcs/tests/serde.test.ts:
```diff
@@ -2,25 +2,27 @@
+        let bcs = new BCS({});
         let u8 = 100;
         let u64 = '18446744073709551615';
         let u128 = '1844674407370955161518446744073709551';
```
I know it just landed, but you might want to include u16, u32, u256 also if you can

Jordan-Mysten commented on sdk/bcs/tests/bcs.config.js:
```diff
@@ -0,0 +1,91 @@
+    {
+        const bcs = new BCS(getRustConfig());
+        let value = [ 'beep', 'boop', 'beep' ];
+        assert_deep_equal(
```
You should use `describe` / `it` / `expect` like the other tests here do.

Jordan-Mysten commented on sdk/bcs/tests/index.js:
```diff
@@ -0,0 +1,4 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+require('./serialization.base.js');
```
You shouldn't need this.

Jordan-Mysten commented on sdk/bcs/package.json:
```diff
@@ -45,6 +45,7 @@
   },
   "devDependencies": {
     "@types/bn.js": "^5.1.0",
+    "pnpm": "^7.13.6",
```
Shouldn't be a dependency.

Jordan-Mysten commented on sdk/bcs/src/index.ts:
```diff
@@ -371,28 +396,134 @@ function ulebDecode(
+   * @param withPrimitives Whether to register primitive types by default
+   */
+  constructor(public schema: BcsConfig, withPrimitives: boolean = true) {
+    this.genericSeparators = schema.genericSeparators || ['<', '>'];
```
If you want to make cloning easier, you could just stash the entire schema instead of assigning the properties into the instance.

Jordan-Mysten commented on sdk/bcs/src/index.ts:
```diff
@@ -371,28 +396,134 @@ function ulebDecode(
+   * @param schema A prepared schema with type definitions
+   * @param withPrimitives Whether to register primitive types by default
+   */
+  constructor(public schema: BcsConfig, withPrimitives: boolean = true) {
```
I always hate boolean arguments, I'd rather this constructor just get a config object that can contain the schema and other configuration.

Jordan-Mysten commented on sdk/bcs/yarn.lock:
```diff
@@ -0,0 +1,919 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
This should be removed.

Jordan-Mysten commented on sdk/bcs/tests/serialization.base.js:
```diff
@@ -0,0 +1,154 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+const { BCS, getSuiMoveConfig } = require('../dist');
```
This should also use tests in the same format as the other tests

666lcz commented on sdk/typescript/src/types/sui-bcs.ts:
```diff
@@ -96,11 +93,9 @@ export type PayAllSuiTx = {
-  .registerVectorType('vector<SuiObjectRef>', 'SuiObjectRef')
   .registerStructType('PayTx', {
     coins: 'vector<SuiObjectRef>',
-    recipients: 'vector<SuiAddress>',
```
you also need to replace line 104 and line 110(on the right)

Jordan-Mysten commented on sdk/typescript/src/types/sui-bcs.ts:
```diff
@@ -1,18 +1,13 @@
 import { Buffer } from 'buffer';
 import { SuiObjectRef } from './objects';
 
+const bcs = new BCS(getSuiMoveConfig());
```
This probably should be exported so that this configured bcs instance can be easily used.

damirka replied:
It was and is exported down in the file.

Jordan-Mysten replied:
Ah sorry I missed that!

Jordan-Mysten commented on sdk/bcs/tests/serde.test.ts:
```diff
@@ -43,8 +59,125 @@ describe('serde', () => {
+    return de;
+}
+
+function assert(cond, tag = '') {
```
Use expect instead of this assert.

Jordan-Mysten commented on sdk/bcs/tests/serde.test.ts:
```diff
@@ -43,8 +59,125 @@ describe('serde', () => {
+    }
+}
+
+function assert_deep_equal(v1, v2, tag = '') {
```
Similarly, you should use expect instead of creating this assert.

666lcz reviewed (approved):
LGTM! Please address Jordan's comments before merging.

damirka commented:
@Jordan-Mysten @666lcz merging this buddy in if tests pass. Addressed your comments, we now use `expect` everywhere, and merged recent u16, u32, u256.

README and npm update will go in a separate PR to speed up current TS SDK development and reduce amount of conflicts.
